### PR TITLE
Symbolic audio generation using Perceiver AR

### PR DESCRIPTION
**Pull request by cstub**
* Implement backend model `SymbolicAudioModel` based on Perceiver AR for symbolic audio modelling
* Provide HuggingFace interface `PerceiverSymbolicAudioModel` and HuggingFace pipeline `symbolic-audio-generation` to use a symbolic audio model for inference
* Implement PyTorch Lightning wrapper to train a symbolic audio model
* Provide HuggingFace dataset implementations for `MaestroV3` and `GiantMIDI-Piano` datasets
* Implement residual dropout in Perceiver core modules
* Add poetry lock check in pre-commit config

### CODE REVIEW

**Comment by cstub** [comment]
Inference examples are missing at the moment.

File: perceiver/data/audio/giantmidi_piano.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+from perceiver.data.audio.symbolic import SymbolicAudioDataModule
+from perceiver.data.audio.utils import download_file, extract_file
+
+
+class GiantMidiPianoDataModule(SymbolicAudioDataModule):
+    def __init__(
+        self,
+        *args: Any,
+        dataset_uri: str = "https://perceiver-io-giant-midi.s3.amazonaws.com/giantmidi-piano.zip",
```
Comment (krasserm) [inline perceiver/data/audio/giantmidi_piano.py:16]: A copy is now available at https://martin-krasser.com/perceiver/data/midi/giantmidi-piano.zip

File: perceiver/data/audio/maestro_v3.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+from perceiver.data.audio.symbolic import SymbolicAudioDataModule
+from perceiver.data.audio.utils import download_file, extract_file
+
+
+class MaestroV3DataModule(SymbolicAudioDataModule):
+    def __init__(
+        self,
+        *args: Any,
+        dataset_uri: str = "https://storage.googleapis.com/magentadata/datasets/maestro/v3.0.0/maestro-v3.0.0-midi.zip",
```
Comment (krasserm) [inline perceiver/data/audio/maestro_v3.py:17]: A copy is now available at https://martin-krasser.com/perceiver/data/midi/maestro-v3.0.0-midi.zip

File: perceiver/data/audio/midi_processor.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+def decode_midi(idx_array, file_path=None) -> pretty_midi.PrettyMIDI:
+    event_sequence = [Event.from_int(idx) for idx in idx_array]
+    snote_seq = _event_seq2snote_seq(event_sequence)
+    note_seq = _merge_note(snote_seq)
+    note_seq.sort(key=lambda x: x.start)
+
+    mid = pretty_midi.PrettyMIDI()
+    # if want to change instument, see https://www.midi.org/specifications/item/gm-level-1-sound-set
+    instument = pretty_midi.Instrument(1, False, "Developed By Yang-Kichang")
+    instument.notes = note_seq
```
Comment (krasserm) [inline perceiver/data/audio/midi_processor.py:254]: Typo `instument` -> `instrument`

File: perceiver/data/audio/midi_processor.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    """Encode a list of midi files using multiple cpu workers."""
+    with Pool(processes=num_workers) as pool:
+        res = list(tqdm(pool.imap(_encode_midi_file, files), total=len(files)))
+        return [r for r in res if r is not None]
+
+
+def _encode_midi_file(file: Path) -> Optional[np.ndarray]:
+    try:
+        midi_file = pretty_midi.PrettyMIDI(str(file))
+        return np.array(encode_midi(midi_file), dtype=np.int32)
```
Comment (krasserm) [inline perceiver/data/audio/midi_processor.py:272]: Why is `np.int32` used here when the data are alter saved as `np.int16`?

File: perceiver/data/audio/symbolic.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    _EXAMPLE_SEPARATOR_INPUT_ID = -1
+    _PAD_INPUT_ID = 388
+    _VOCAB_SIZE = 389
+
+    def __init__(
+        self,
+        dataset_dir: str,
+        max_seq_len: int,
+        min_seq_len: Optional[int] = None,
+        padding_side: Optional[str] = "left",
```
Comment (krasserm) [inline perceiver/data/audio/symbolic.py:27]: Use `None` as default for an optional parameter.

File: perceiver/data/audio/symbolic.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        batch_size: int = 16,
+        num_workers: int = 1,
+        preproc_workers: int = multiprocessing.cpu_count(),
```
Comment (krasserm) [inline perceiver/data/audio/symbolic.py:30]: Make optional and initialize to `num_workers` to be consistent with other data modules.

File: perceiver/data/audio/symbolic.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    @property
+    def max_seq_len(self):
+        return self.hparams.max_seq_len
+
+    @property
+    @functools.lru_cache(maxsize=1)
+    def preproc_dir(self) -> Path:
+        return Path(self.hparams.dataset_dir) / "preproc" / self.preproc_dir_hash()
+
+    def preproc_dir_hash(self) -> str:
```
Comment (krasserm) [inline perceiver/data/audio/symbolic.py:63]: I think we can remove this and the implementations in subclasses. It is safe to assume that different `dataset_uri`s will always host the same dataset file.

File: perceiver/data/audio/symbolic.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        if self._separator_input_id not in sample:
+            example = sample
+        else:
+            examples = list(
+                torch.tensor_split(sample, torch.where(sample == self._separator_input_id)[0])
+            )  # split at example separators
+            examples = sorted(examples, key=lambda x: len(x), reverse=True)  # order by longest sample
+
+            example = examples[0]
+            example = example[example != self._separator_input_id]
```
Comment (krasserm) [inline perceiver/data/audio/symbolic.py:166]: Why is this needed? Doesn't `example != self._separator_input_id` always evaluate to a tensor filled with to `True` here?

File: perceiver/data/audio/utils.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
+import zipfile
+from pathlib import Path
+
+import requests
+
+
+def download_file(uri: str, target_file: Path):
```
Comment (krasserm) [inline perceiver/data/audio/utils.py:7]: Isn't this functionality part of one of the dependencies we are using?

File: perceiver/model/core/modules.py [hunk 10]
```diff
@@ -376,6 +380,7 @@ def __init__(
         first_self_attention_block_shared: bool = True,
         self_attention_widening_factor: int = 1,
         dropout: float = 0.0,
+        residual_dropout: float = 0.0,
```
Comment (krasserm) [inline perceiver/model/core/modules.py:383]: Please add this parameter to the API docs (only if these already exist) and extend the description of `dropout` to contain "post-attention dropout".

File: pyproject.toml [hunk 11]
```diff
@@ -69,6 +70,7 @@ termcolor = "2.0.1"
 [tool.poetry.extras]
 text = ["datasets", "tokenizers", "transformers"]
 vision = ["datasets", "torchvision", "opencv-python"]
+audio = ["datasets", "pretty_midi", "transformers"]
```
Comment (krasserm) [inline pyproject.toml:73]: Please update installation instructions in README.md to contain this additional extra.

File: tests/conftest.py [hunk 12]
```diff
@@ -1,7 +1,11 @@
 import tempfile
+from pathlib import Path
```
Comment (krasserm) [inline tests/conftest.py:2]: I understand that using `pathlib` is a better choice than `os.path` nowadays but I'd prefer to have a consistent usage in this project. Does anything speak using `os.path` in this pull request and switch the entire project to `pathlib` in a separate pull request?

**Review by krasserm — COMMENTED** [review]
Very good contribution, thanks @cstub! I only had some minor comments. Please send me the training checkpoint through another channel so that I can upload it to the FTP server.

File: perceiver/data/audio/symbolic.py (hunk 8 above)
Comment (cstub) [reply perceiver/data/audio/symbolic.py:166]: The `sample` is split into individual `examples` at each occurrence of the `self._separator_input_id` using `torch.tensor_split`. `torch.tensor_split` preserves the occurrences of each `self._separator_id` which is why they are removed here.

File: .pre-commit-config.yaml [hunk 13]
```diff
@@ -49,3 +49,10 @@ repos:
     rev: 4.0.1
     hooks:
       - id: flake8
+
+  - repo: https://github.com/python-poetry/poetry
+    rev: 1.2.1
+    hooks:
+      - id: poetry-check
+      - id: poetry-lock
+        args: [ "--check" ]
```
[the remark this answers is not in the document]
Comment (cstub) [reply .pre-commit-config.yaml:58]: No specific reason other than it is good practice, avoids the possibility of inconsistencies and is cheap to run.

File: docs/pretrained-models.md [hunk 14]
```diff
@@ -139,3 +139,20 @@ processor = AutoImageProcessor.from_pretrained(repo_id)
 
 classifier_pipeline = pipeline("image-classification", model=repo_id)
 ```
+
+### [krasserm/perceiver-ar-symbolic-audio](https://huggingface.co/krasserm/perceiver-ar-symbolic-audio)
+
+A medium Perceiver AR audio model trained on the [GiantMIDI-Piano](https://github.com/bytedance/GiantMIDI-Piano) dataset
+in [this training example](training-examples.md#giantmidi-piano).
+The model has 134M parameters and was trained for 27 epochs (153M tokens per epoch).
```
[the remark this answers is not in the document]
Comment (cstub) [reply docs/pretrained-models.md:147]: I plan to conduct some more experiments fine-tuning on Maestro v3 using the latest checkpoint but previous experiments in this direction showed a degradation of results compared to the pretrained model.

File: perceiver/data/audio/midi_processor.py (hunk 4 above)
Comment (cstub) [reply perceiver/data/audio/midi_processor.py:272]: good catch

File: perceiver/data/audio/symbolic.py (hunk 5 above)
Comment (cstub) [reply perceiver/data/audio/symbolic.py:27]: I think in this case its preferable to drop the `Optional` type

File: docs/pretrained-models.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (krasserm) [reply docs/pretrained-models.md:147]: Understood, then fine having the pretrained checkpoint only.

File: perceiver/data/audio/symbolic.py (hunk 5 above)
Comment (krasserm) [reply perceiver/data/audio/symbolic.py:27]: All other modules have `Optional` with default `None`. Shouldn't we prefer consistency now and do a library-wide change later if needed?

File: perceiver/data/audio/symbolic.py (hunk 8 above)
Comment (krasserm) [reply perceiver/data/audio/symbolic.py:166]: `self._separator_input_id` doesn't occur in the first split element, only in elements 2..n. This is why you can omit this line. Or am I missing something?

Comment (krasserm) [reply perceiver/data/audio/symbolic.py:166]: Hmmm ... do I miss the special case when the separator is the very first element in the entire sequence? Can this happen at all?

Comment (cstub) [reply perceiver/data/audio/symbolic.py:166]: The split elements are sorted by length, so the first split element after the sort operation might contain the separator_id.

Comment (krasserm) [reply perceiver/data/audio/symbolic.py:166]: :+1: Ah right, this is what I was missing ...

File: perceiver/data/audio/symbolic.py (hunk 5 above)
Comment (cstub) [reply perceiver/data/audio/symbolic.py:27]: In other modules (e.g. `TextDataModule`) it is acceptable to set the `padding_side` to `None`, as the default of the tokenizer is applied in this case. In this data module, we don't use a tokenizer with a set default, so we have to define the default value. We can set the default value manually in the constructor or in the `SymbolicAudioCollator` but I think setting the default here and consequently preserving the value in the `hparams` would be preferrable.

Comment (krasserm) [reply perceiver/data/audio/symbolic.py:27]: :+1: Good point, I agree

File: perceiver/data/audio/symbolic.py (hunk 6 above)
Comment (cstub) [reply perceiver/data/audio/symbolic.py:30]: `preproc_workers` should be set to a high value as it controls the number of workers encoding the midi files, whereas there is no benefit of setting `num_workers > 1` as each dataset split is read from a single file.
How about we remove the `num_workers` option altogether?

File: tests/conftest.py (hunk 12 above)
Comment (cstub) [reply tests/conftest.py:2]: I vastly prefer the usage of `pathlib` (for numerous reasons) but if consistency is a priority we can switch to `os.path` for the time being.

File: README.md [hunk 15]
```diff
@@ -136,6 +138,110 @@ Here is a side-by-side comparison of the input and output video:
...
+output = audio_generator(prompt, max_new_tokens=64, num_latents=1, do_sample=True, top_p=0.95, temperature=1.0, render=True)
+
+with open("generated_audio.wav", "wb") as f:
+    f.write(output["generated_audio_wav"])
+```
+
+##### Audio samples
+
+The following (hand-picked) audio samples were generated using various prompts from the validation subset of
+the [GiantMIDI-Piano](https://github.com/bytedance/GiantMIDI-Piano) dataset. The input prompts are
```
Comment (krasserm) [inline README.md:166]: Remove "(hand-picked)"?

Comment (krasserm) [reply README.md:166]: Replace "input prompt" with "prompt"?

File: README.md [hunk 16]
```diff
@@ -136,6 +138,110 @@ Here is a side-by-side comparison of the input and output video:
...
+    <th>Temperature</th>
+    <th>Prefix length</th>
+    <th>Latents</th>
+  </tr>
+  <tr>
+    <td>
+      <audio controls>
+        <source src="https://martin-krasser.com/perceiver/data/midi/01_nehrlich_continuation.wav" type="audio/wav">
+        Your browser does not support the audio element.
+      </audio>
```
Comment (krasserm) [inline README.md:183]: My Browser shows the error message. What am I missing?

File: examples/inference.ipynb [hunk 17]
```diff
@@ -1879,6 +1901,262 @@
         "    print_prompt_and_generated(p, g)\n",
         "    print()"
       ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "### Perceiver AR symbolic audio modeling\n",
+        "\n",
+        "The following section demonstrates how to use a Perceiver AR audio model to generate symbolic ([MIDI](https://en.wikipedia.org/wiki/MIDI)) audio data. The model has `134M` parameters and was trained with an input context of `4096` prefix tokens and `2048` latent tokens for `27` epochs (`157M` tokens per epoch) on the [GiantMIDI-Piano](https://github.com/bytedance/GiantMIDI-Piano) dataset. \n",
```
Comment (krasserm) [inline examples/inference.ipynb:1910]: - ~~The following~~ This section
- a~~n input~~ context

For all further occurrences of "input context" or "input prompt" I'd remove "input"

File: examples/inference.ipynb [hunk 18]
```diff
@@ -1879,6 +1901,262 @@
...
+        "\n",
+        "The model accepts tokenized audio symbols extracted from [MIDI](https://en.wikipedia.org/wiki/MIDI) files as input. MIDI files are tokenized using the approach presented in the [Perceiver AR paper](https://arxiv.org/pdf/2202.07765.pdf). The tokenization method is described in detail in Section A.2 of [Huang et al (2019)](https://arxiv.org/abs/1809.04281) and the implementation used in this project is taken from the [Midi Neural Processor](https://github.com/jason9693/midi-neural-processor/blob/bea0dc612b7f687f964d0f6d54d1dbf117ae1307/processor.py).\n",
+        "\n",
+        "To generate audio sequences a 🤗 pipeline can be used. This pipeline is supplied with one or multiple MIDI files of type `pretty_midi.PrettyMIDI` acting as prompts. The pipeline returns the generated sequences either as `pretty_midi.PrettyMIDI` objects (using the option `render=False`) or as a `bytes` array containing the rendered [WAV](https://en.wikipedia.org/wiki/WAV) audio content (using `render=True`). Note that the utility [fluidsynth](https://www.fluidsynth.org/) must be installed on your system to render a MIDI file to the WAV audio format."
```
Comment (krasserm) [inline examples/inference.ipynb:1914]: Note that ~~the utility~~ fluidsynth must be installed ~~on your system~~ to render

(this should still be clear enough)

File: examples/inference.ipynb [hunk 19]
```diff
@@ -1879,6 +1901,262 @@
...
+        "id": "i9J3QtxmV_85"
+      },
+      "id": "i9J3QtxmV_85",
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "This example demonstrate how to generate an audio sequence without adding the input prompt. It `4096` tokens from the input prompt to generate `1024` new tokens. The initial number of latent positions is set to `1` corresponding to the last `1` tokens of the prompt sequence. All other prompt tokens are used as prefix tokens. Only the generated sequence is returned (`return_full_audio=False`)."
```
Comment (krasserm) [inline examples/inference.ipynb:2126]: - without ~~adding~~ returning the input prompt
- last 1 token~~s~~

File: examples/inference.ipynb [hunk 20]
```diff
@@ -1879,6 +1901,262 @@
...
+        "id": "Idbid6snVzQo"
+      },
+      "id": "Idbid6snVzQo",
+      "execution_count": 45,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "The following example demonstrates how to create a continuation of an input prompt. It uses `512` tokens from the input prompt to generate `1024` new tokens. The initial number of latent positions is set to `1` corresponding to the last `1` tokens of the prompt sequence. All other prompt tokens are used as prefix tokens. The generated sequence is appended to the input prompt (`return_full_audio=True`) and the output is return as a binary wav sequence. The generated sequence starts at timestamp `14s`."
```
Comment (krasserm) [inline examples/inference.ipynb:2088]: output is return<ins>ed</ins> ~~as a binary wav sequence~~ in binary wav format.

**Review by krasserm — COMMENTED** [review]
Awesome additions @cstub, only had some really minor comments. Do you also plan to add audio samples to the inference notebook? If not, please link to the corresponding section in the model card on the Hugging Face Hub.

File: README.md (hunk 16 above)
Comment (cstub) [reply README.md:183]: Apparently audio tags are not rendered correctly on GitHub README pages. How about we add a link to the HF model card instead?
